certbot/error_handler.py

```diff
@@ -20,43 +21,64 @@
 
-    with ErrorHandler(cleanup_func):
+    handler = ErrorHandler(cleanup1_func, *cleanup1_args, **cleanup1_kwargs)
+    handler.register(cleanup2_func, *cleanup2_args, **cleanup2_kwargs)
+
+    with handler:
         do_something()
 
-    If an exception is raised out of do_something, cleanup_func will be
-    called. The exception is not caught by the ErrorHandler. Similarly,
-    if a signal is encountered, cleanup_func is called followed by the
-    previously registered signal handler.
+    Or for one cleanup function:
+
+    with ErrorHandler(func, args, kwargs):
+        do_something()
 
-    Every registered function is attempted to be run to completion
-    exactly once. If a registered function raises an exception, it is
-    logged and the next function is called. If a (different) handled
-    signal occurs while calling a registered function, it is attempted
-    to be called again by the next signal handler.
+    If an exception is raised out of do_something, the cleanup functions will
+    be called in last in first out order. Then the exception is raised.
+    Similarly, if a signal is encountered, the cleanup functions are called
+    followed by the previously received signal handler.
+
+    Each registered cleanup function is called exactly once. If a registered
+    function raises an exception, it is logged and the next function is called.
+    Signals received while the registered functions are executing are
+    deferred until they finish.
 
     """
     def __init__(self, func=None, *args, **kwargs):
+        self.body_executed = False
         self.funcs = []
         self.prev_handlers = {}
+        self.received_signals = []
         if func is not None:
             self.register(func, *args, **kwargs)
 
     def __enter__(self):
-        self.set_signal_handlers()
+        self.body_executed = False
+        self._set_signal_handlers()
 
     def __exit__(self, exec_type, exec_value, trace):
-        # SystemExit is ignored to properly handle forks that don't exec
-        if exec_type not in (None, SystemExit):
-            logger.debug("Encountered exception:\n%s", "".join(
-                traceback.format_exception(exec_type, exec_value, trace)))
-            self.call_registered()
-        self.reset_signal_handlers()
+        try:
+            self.body_executed = True
+            retval = False
+            # SystemExit is ignored to properly handle forks that don't exec
+            if exec_type in (None, SystemExit):
+                return retval
+            elif exec_type is errors.SignalExit:
+                logger.debug("Encountered signals: %s", self.received_signals)
+                retval = True
+            else:
+                logger.debug("Encountered exception:\n%s", "".join(
+                    traceback.format_exception(exec_type, exec_value, trace)))
+
+            self._call_registered()
+            return retval
+        finally:
```

Comment:
Question: What is the purpose of the `try`/`finally`? Where are we worried about an exception being raised? Is it in case we get a signal before the `self.body_executed = True` line?